### PR TITLE
[MultiTenant] Added CPU throttling feature

### PR DESCRIPTION
**Pull request by luchsh**
Summary: ported cgroup-based cpu throttling facilities

Test Plan: jdk/test/multi-tenant

Reviewed-by: yuleil, superajun-wsj

Issue: https://github.com/alibaba/dragonwell8/issues/84

### CODE REVIEW

File: src/linux/native/jgroup/com/alibaba/tenant/jgroup.c [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+
+    return res;
+}
+
+// Detect cgroup mountpoints
+static int
+detect_cgroup_mount_points() {
+    int res = 0;
+	struct mntent *ent = NULL;
+	char mntent_buffer[4 * MAX_PATH_LEN];
```
Comment (superajun-wsj) [inline src/linux/native/jgroup/com/alibaba/tenant/jgroup.c:169]: alignment

File: src/linux/native/jgroup/com/alibaba/tenant/jgroup.c [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+    int cg_index = 0;
+    FILE* proc_mounts = NULL;
+    struct mntent* temp_ent = NULL;
+
+	temp_ent = (struct mntent *) malloc(sizeof(struct mntent));
```
Comment (superajun-wsj) [inline src/linux/native/jgroup/com/alibaba/tenant/jgroup.c:174]: alignment

File: src/linux/native/jgroup/com/alibaba/tenant/jgroup.c [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+                snprintf(path_buf, MAX_PATH_LEN, "%s/%s", cg_mount_points[cg_index], path_str);
+
+                // if exists, just reuse
+                if (stat(path_buf, &st) == 0 && S_ISDIR(st.st_mode)) {
+                    JGROUP_INFO("Reuse existing directory %s", path_buf);
+                    continue;
+                }
+
+                res = mkdir(path_buf,
+                            S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH);  // modes are copied from libcg
```
Comment (superajun-wsj) [inline src/linux/native/jgroup/com/alibaba/tenant/jgroup.c:341]: should it be placed at the same line with 'mkdir' call.

File: src/linux/native/jgroup/com/alibaba/tenant/jgroup.c [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+                int sz = fprintf(fp_tasks, "%d", (int)tid);
+                fclose(fp_tasks);
+
+                if (sz > 0) {
+                    JGROUP_INFO("PID %d has been writen to %s", (int)tid, path_buf);
+                } else {
+                    JGROUP_ERROR("Failed to write to %s", path_buf);
+                }
+
+
```
Comment (superajun-wsj) [inline src/linux/native/jgroup/com/alibaba/tenant/jgroup.c:388]: remove these space lines.

File: src/linux/native/jgroup/com/alibaba/tenant/jgroup.c (hunk 1 above)
Comment (luchsh) [reply src/linux/native/jgroup/com/alibaba/tenant/jgroup.c:169]: done.
there were mixed tabs and spaces, just removed tabs.

File: src/linux/native/jgroup/com/alibaba/tenant/jgroup.c (hunk 3 above)
Comment (luchsh) [reply src/linux/native/jgroup/com/alibaba/tenant/jgroup.c:341]: done

File: src/linux/native/jgroup/com/alibaba/tenant/jgroup.c (hunk 2 above)
Comment (luchsh) [reply src/linux/native/jgroup/com/alibaba/tenant/jgroup.c:174]: done

File: src/share/classes/com/alibaba/tenant/TenantConfiguration.java [hunk 5]
```diff
@@ -21,16 +21,23 @@
 
 package com.alibaba.tenant;
 
+import com.alibaba.rcm.Constraint;
+import com.alibaba.rcm.ResourceType;
```
Comment (superajun-wsj) [inline src/share/classes/com/alibaba/tenant/TenantConfiguration.java:25]: the two import packages are duplicated

File: src/share/classes/com/alibaba/tenant/TenantResourceType.java [hunk 6]
```diff
@@ -22,12 +22,153 @@
...
+        }
+    };
+
+    /**
+     * Throttle CPU resource with cgroup controller: cpuset.cpus.
+     * Expecting N constraint value which represent cores described in cpuset.cpus.
+     * e.g. {@code TenantResourceType.CPUSET_CPUS.newConstraint(LongStream.range(0, 16).toArray()) }
+     */
+    static final ResourceType CPUSET_CPUS = new TenantResourceType("CPUSET_CPUS",
+            true) {
```
Comment (superajun-wsj) [inline src/share/classes/com/alibaba/tenant/TenantResourceType.java:68]: make the two parameters of constructor method 'TenantResourceType' in the same line

File: test/multi-tenant/TestNoJGroupInit.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+import java.lang.reflect.*;
```
Comment (superajun-wsj) [inline test/multi-tenant/TestNoJGroupInit.java:21]: we only use one method 'Field' of reflect package, so should not use reflect.*

**Pull request merged**